[AndroidTV] Unable to establish baseline for android tv

VIVIDUS: 0.3.9-SNAPSHOT

```shell
Caused by: org.openqa.selenium.WebDriverException: An unknown server-side error occurred while processing the command. Original error: Failed to parse status bar information.
Build info: version: '4.0.0-alpha-1', revision: 'd1d3728cae', time: '2019-04-24T13:42:21'
System info: host: '***', ip: 'fe80:0:0:0:c7a:fe51:14f4:e5bb%en0', os.name: 'Mac OS X', os.arch: 'x86_64', os.version: '10.15.6', java.version: '16.0.2'
Driver info: io.appium.java_client.android.AndroidDriver
Capabilities {app: .., appPackage: *** appiumVersion: 1.20.2, automationName: UiAutomator2, databaseEnabled: false, desired: {app: ..., appiumVersion: 1.20.2, automationName: UiAutomator2, platformName: android, udid: ***}, deviceApiLevel: 28, deviceManufacturer: Xiaomi, deviceModel: MIBOX4, deviceName: 18554282254222, deviceScreenDensity: 320, deviceScreenSize: 1920x1080, deviceUDID: ****, javascriptEnabled: true, locationContextEnabled: false, networkConnectionEnabled: true, pixelRatio: 2, platform: LINUX, platformName: Android, platformVersion: 9, statBarHeight: 0, takesScreenshot: true, udid: ****, viewportRect: {height: 1080, left: 0, top: 0, width: 1920}, warnings: {}, webStorageEnabled: false}
Session ID: 909df899-b70d-4de8-95fd-186d1fad4a47
	at jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method) ~[?:?]
	at jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:78) ~[?:?]
	at jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45) ~[?:?]
	at java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:499) ~[?:?]
	at java.lang.reflect.Constructor.newInstance(Constructor.java:480) ~[?:?]
	at org.openqa.selenium.remote.codec.w3c.W3CHttpResponseCodec.createException(W3CHttpResponseCodec.java:190) ~[selenium-remote-driver-4.0.0-alpha-1.jar:?]
	at org.openqa.selenium.remote.codec.w3c.W3CHttpResponseCodec.decode(W3CHttpResponseCodec.java:124) ~[selenium-remote-driver-4.0.0-alpha-1.jar:?]
	at org.openqa.selenium.remote.codec.w3c.W3CHttpResponseCodec.decode(W3CHttpResponseCodec.java:51) ~[selenium-remote-driver-4.0.0-alpha-1.jar:?]
	at org.openqa.selenium.remote.HttpCommandExecutor.execute(HttpCommandExecutor.java:161) ~[selenium-remote-driver-4.0.0-alpha-1.jar:?]
	at io.appium.java_client.remote.AppiumCommandExecutor.execute(AppiumCommandExecutor.java:239) ~[java-client-7.1.0.jar:?]
	at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:577) ~[selenium-remote-driver-4.0.0-alpha-1.jar:?]
	at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:634) ~[selenium-remote-driver-4.0.0-alpha-1.jar:?]
	at io.appium.java_client.DefaultGenericMobileDriver.execute(DefaultGenericMobileDriver.java:41) ~[java-client-7.1.0.jar:?]
	at io.appium.java_client.AppiumDriver.execute(AppiumDriver.java:1) ~[java-client-7.1.0.jar:?]
	at io.appium.java_client.android.AndroidDriver.execute(AndroidDriver.java:1) ~[java-client-7.1.0.jar:?]
	at io.appium.java_client.CommandExecutionHelper.execute(CommandExecutionHelper.java:27) ~[java-client-7.1.0.jar:?]
	at org.vividus.selenium.mobileapp.MobileAppWebDriverManager.getAndroidStatusBar(MobileAppWebDriverManager.java:62) ~[vividus-plugin-mobile-app-0.3.9-SNAPSHOT.jar:0.3.9-SNAPSHOT]
	at org.vividus.selenium.mobileapp.MobileAppWebDriverManager.getStatusBarSize(MobileAppWebDriverManager.java:55) ~[vividus-plugin-mobile-app-0.3.9-SNAPSHOT.jar:0.3.9-SNAPSHOT]
	at org.vividus.selenium.mobileapp.screenshot.MobileAppAshotFactory.create(MobileAppAshotFactory.java:56) ~[vividus-plugin-mobile-app-0.3.9-SNAPSHOT.jar:0.3.9-SNAPSHOT]
	at org.vividus.selenium.screenshot.AbstractScreenshotTaker.crateAshot(AbstractScreenshotTaker.java:124) ~[vividus-extension-selenium-0.3.9-SNAPSHOT.jar:0.3.9-SNAPSHOT]
	at org.vividus.selenium.screenshot.AbstractScreenshotTaker.takeAshotScreenshot(AbstractScreenshotTaker.java:109) ~[vividus-extension-selenium-0.3.9-SNAPSHOT.jar:0.3.9-SNAPSHOT]
	at org.vividus.visual.screenshot.AshotScreenshotProvider.take(AshotScreenshotProvider.java:64) ~[vividus-extension-visual-testing-0.3.9-SNAPSHOT.jar:0.3.9-SNAPSHOT]
	at org.vividus.visual.engine.VisualTestingEngine.getCheckpointScreenshot(VisualTestingEngine.java:73) ~[vividus-plugin-visual-0.3.9-SNAPSHOT.jar:0.3.9-SNAPSHOT]
	at org.vividus.visual.engine.VisualTestingEngine.establish(VisualTestingEngine.java:65) ~[vividus-plugin-visual-0.3.9-SNAPSHOT.jar:0.3.9-SNAPSHOT]
	at org.vividus.visual.VisualSteps.lambda$performVisualAction$2(VisualSteps.java:98) ~[vividus-plugin-visual-0.3.9-SNAPSHOT.jar:0.3.9-SNAPSHOT]
	at org.vividus.visual.bdd.AbstractVisualSteps.execute(AbstractVisualSteps.java:53) ~[vividus-extension-visual-testing-0.3.9-SNAPSHOT.jar:0.3.9-SNAPSHOT]
	at org.vividus.visual.VisualSteps.performVisualAction(VisualSteps.java:94) ~[vividus-plugin-visual-0.3.9-SNAPSHOT.jar:0.3.9-SNAPSHOT]
	at org.vividus.visual.VisualSteps.runVisualTests(VisualSteps.java:73) ~[vividus-plugin-visual-0.3.9-SNAPSHOT.jar:0.3.9-SNAPSHOT]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:78) ~[?:?]
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:567) ~[?:?]
	at org.jbehave.core.steps.StepCreator$ParametrisedStep.perform(StepCreator.java:861) ~[jbehave-core-5.0.0-alpha.13.jar:?]
	... 18 more
```

--- a/vividus-plugin-mobile-app/src/main/java/org/vividus/selenium/mobileapp/MobileAppWebDriverManager.java
+++ b/vividus-plugin-mobile-app/src/main/java/org/vividus/selenium/mobileapp/MobileAppWebDriverManager.java
@@ -31,6 +31,7 @@ import com.google.common.base.Suppliers;
 
 import org.openqa.selenium.OutputType;
 import org.openqa.selenium.TakesScreenshot;
+import org.openqa.selenium.WebDriverException;
 import org.vividus.selenium.IWebDriverProvider;
 import org.vividus.selenium.manager.GenericWebDriverManager;
 import org.vividus.selenium.manager.IWebDriverManagerContext;
@@ -52,21 +53,31 @@ public class MobileAppWebDriverManager extends GenericWebDriverManager
 
     public int getStatusBarSize()
     {
-        return  super.isIOSNativeApp() ? getIOsStatusBar() : getAndroidStatusBar();
+        return  super.isIOSNativeApp() ? getStatBarHeight() : getAndroidStatusBar();
     }
 
     private int getAndroidStatusBar()
     {
-        HasAndroidDeviceDetails details = getWebDriverProvider().getUnwrapped(HasAndroidDeviceDetails.class);
-        // https://github.com/appium/java-client/commit/9020174c578bed5e03c24b43c2c5ba590f663201
-        Map<String, Map<String, Object>>  systemBars = CommandExecutionHelper.execute(details,
-                AndroidMobileCommandHelper.getSystemBarsCommand());
-        return Optional.ofNullable(systemBars)
-                       .map(b -> b.get("statusBar"))
-                       .map(sb -> sb.get("height"))
-                       .map(Long.class::cast)
-                       .map(Long::intValue)
-                       .orElse(0);
+        try
+        {
+            HasAndroidDeviceDetails details = getWebDriverProvider().getUnwrapped(HasAndroidDeviceDetails.class);
+            // https://github.com/appium/java-client/commit/9020174c578bed5e03c24b43c2c5ba590f663201
+            Map<String, Map<String, Object>>  systemBars = CommandExecutionHelper.execute(details,
+                    AndroidMobileCommandHelper.getSystemBarsCommand());
+            return Optional.ofNullable(systemBars)
+                           .map(b -> b.get("statusBar"))
+                           .map(sb -> sb.get("height"))
+                           .map(Long.class::cast)
+                           .map(Long::intValue)
+                           .orElse(0);
+        }
+        catch (WebDriverException e)
+        {
+            // The workaround is for Android TV. It is not clear if any of the Android TVs could have a status bar
+            // at all, but the session capabilities contains `statBarHeight`, and to be on the safe side fall-back code
+            // was added to get the status bar height in case of exception.
+            return getStatBarHeight();
+        }
     }
 
     public double getDpr()
@@ -89,7 +100,7 @@ public class MobileAppWebDriverManager extends GenericWebDriverManager
         }
     }
 
-    private int getIOsStatusBar()
+    private int getStatBarHeight()
     {
         HasSessionDetails details = getWebDriverProvider().getUnwrapped(HasSessionDetails.class);
         return ((Long) details.getSessionDetail("statBarHeight")).intValue();

--- a/vividus-plugin-mobile-app/src/test/java/org/vividus/selenium/mobileapp/MobileAppWebDriverManagerTests.java
+++ b/vividus-plugin-mobile-app/src/test/java/org/vividus/selenium/mobileapp/MobileAppWebDriverManagerTests.java
@@ -31,8 +31,6 @@ import java.util.Map;
 
 import javax.imageio.ImageIO;
 
-import com.google.common.collect.ImmutableMap;
-
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.extension.ExtendWith;
 import org.mockito.InjectMocks;
@@ -46,6 +44,7 @@ import org.openqa.selenium.HasCapabilities;
 import org.openqa.selenium.OutputType;
 import org.openqa.selenium.TakesScreenshot;
 import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.WebDriverException;
 import org.openqa.selenium.remote.CapabilityType;
 import org.openqa.selenium.remote.Response;
 import org.vividus.selenium.IWebDriverProvider;
@@ -58,6 +57,10 @@ import io.appium.java_client.remote.MobilePlatform;
 @ExtendWith(MockitoExtension.class)
 class MobileAppWebDriverManagerTests
 {
+    private static final String GET_SYSTEM_BARS = "getSystemBars";
+
+    private static final String STAT_BAR_HEIGHT = "statBarHeight";
+
     private static final byte[] IMAGE = new byte[] { (byte) 0x89, 0x50, 0x4E, 0x47, 0x0D, 0x0A,
             0x1A, 0x0A, 0x00, 0x00, 0x00, 0x0D, 0x49, 0x48, 0x44, 0x52, 0x00, 0x00, 0x00, 0x01, 0x00, 0x00, 0x00,
             0x01, 0x08, 0x06, 0x00, 0x00, 0x00, 0x1F, 0x15, (byte) 0xC4, (byte) 0x89, 0x00, 0x00, 0x00, 0x0B, 0x49,
@@ -76,8 +79,21 @@ class MobileAppWebDriverManagerTests
         when(webDriverProvider.getUnwrapped(HasAndroidDeviceDetails.class)).thenReturn(hasAndroidDeviceDetails);
         Response response = mock(Response.class);
         when(response.getValue()).thenReturn(Map.of("statusBar", Map.of("height", 100L)));
-        when(hasAndroidDeviceDetails.execute("getSystemBars", ImmutableMap.of())).thenReturn(response);
+        when(hasAndroidDeviceDetails.execute(GET_SYSTEM_BARS, Map.of())).thenReturn(response);
         assertEquals(100, driverManager.getStatusBarSize());
+    }
+
+    @Test
+    void shouldTryToGetStatBarHeightInCaseOfWebDriverException()
+    {
+        HasAndroidDeviceDetails hasAndroidDeviceDetails = mock(HasAndroidDeviceDetails.class,
+                withSettings().extraInterfaces(ExecutesMethod.class));
+        when(webDriverProvider.getUnwrapped(HasAndroidDeviceDetails.class)).thenReturn(hasAndroidDeviceDetails);
+        when(hasAndroidDeviceDetails.execute(GET_SYSTEM_BARS, Map.of())).thenThrow(new WebDriverException());
+        HasSessionDetails details = mock(HasSessionDetails.class);
+        when(webDriverProvider.getUnwrapped(HasSessionDetails.class)).thenReturn(details);
+        when(details.getSessionDetail(STAT_BAR_HEIGHT)).thenReturn(101L);
+        assertEquals(101, driverManager.getStatusBarSize());
     }
 
     @Test
@@ -91,7 +107,7 @@ class MobileAppWebDriverManagerTests
         when(capabilitiesMock.getCapability(CapabilityType.PLATFORM_NAME)).thenReturn(MobilePlatform.IOS);
         HasSessionDetails details = mock(HasSessionDetails.class);
         when(webDriverProvider.getUnwrapped(HasSessionDetails.class)).thenReturn(details);
-        when(details.getSessionDetail("statBarHeight")).thenReturn(102L);
+        when(details.getSessionDetail(STAT_BAR_HEIGHT)).thenReturn(102L);
         assertEquals(102, driverManager.getStatusBarSize());
     }
 

